File: bot-api/dotnet/test/src/InitialPositionMapperTest.cs
using NUnit.Framework;
using Robocode.TankRoyale.BotApi;
using Robocode.TankRoyale.BotApi.Mapper;

namespace Robocode.TankRoyale.BotApi.Tests;

public class InitialPositionMapperTest
{
    [Test]
    [Category("VAL")]
    [Property("ID", "TR-API-VAL-004")]
    public void GivenInitialPosition_whenMapped_thenSchemaValuesArePreserved()
    {
        // Arrange
        var ip = InitialPosition.FromString("11, 22, 33");

        // Act
        var schemaIp = InitialPositionMapper.Map(ip);

        // Assert
        Assert.That(schemaIp, Is.Not.Null);
        Assert.That(schemaIp.X, Is.EqualTo(11.0));
        Assert.That(schemaIp.Y, Is.EqualTo(22.0));
        Assert.That(schemaIp.Direction, Is.EqualTo(33.0));

        // Round-trip back via values
        var ip2 = InitialPosition.FromString($"{schemaIp.X}, {schemaIp.Y}, {schemaIp.Direction}");
        Assert.That(ip2, Is.Not.Null);
        Assert.That(ip2.X, Is.EqualTo(ip.X));
        Assert.That(ip2.Y, Is.EqualTo(ip.Y));
        Assert.That(ip2.Direction, Is.EqualTo(ip.Direction));
    }

    [Test]
    [Category("VAL")]
    [Property("ID", "TR-API-VAL-004")]
    public void GivenNullOrPartialInitialPosition_whenMapped_thenNullOrPartialSchemaIsReturned()
    {
        // Null source -> null schema
        var schemaNull = InitialPositionMapper.Map(null);
        Assert.That(schemaNull, Is.Null);

        // Partial values preserved (x only)
        var ipXOnly = InitialPosition.FromString("50");
        var schemaXOnly = InitialPositionMapper.Map(ipXOnly);
        Assert.That(schemaXOnly, Is.Not.Null);
        Assert.That(schemaXOnly.X, Is.EqualTo(50.0));
        Assert.That(schemaXOnly.Y, Is.Null);
        Assert.That(schemaXOnly.Direction, Is.Null);
    }
}

File: bot-api/java/src/test/java/dev/robocode/tankroyale/botapi/mapper/InitialPositionMapperTest.java
package dev.robocode.tankroyale.botapi.mapper;

import dev.robocode.tankroyale.botapi.InitialPosition;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class InitialPositionMapperTest {

    @Test
    @DisplayName("TR-API-VAL-004 InitialPosition mapping round-trip: mapper preserves values")
    @Tag("VAL")
    @Tag("TR-API-VAL-004")
    void test_TR_API_VAL_004_mapper_preserves_values() {
        // Arrange
        var ip = InitialPosition.fromString("11, 22, 33");

        // Act
        var schemaIp = InitialPositionMapper.map(ip);

        // Assert
        assertThat(schemaIp).isNotNull();
        assertThat(schemaIp.getX()).isEqualTo(11.0);
        assertThat(schemaIp.getY()).isEqualTo(22.0);
        assertThat(schemaIp.getDirection()).isEqualTo(33.0);

        // Round-trip back via values to InitialPosition (like-for-like)
        var ip2 = InitialPosition.fromString(
                schemaIp.getX() + ", " + schemaIp.getY() + ", " + schemaIp.getDirection());
        assertThat(ip2).isNotNull();
        assertThat(ip2.getX()).isEqualTo(ip.getX());
        assertThat(ip2.getY()).isEqualTo(ip.getY());
        assertThat(ip2.getDirection()).isEqualTo(ip.getDirection());
    }

    @Test
    @DisplayName("TR-API-VAL-004 InitialPosition mapping handles null and partial values")
    @Tag("VAL")
    @Tag("TR-API-VAL-004")
    void test_TR_API_VAL_004_mapper_handles_nulls() {
        // Null source -> null schema
        var schemaNull = InitialPositionMapper.map(null);
        assertThat(schemaNull).isNull();

        // Partial values preserved
        var ipXOnly = InitialPosition.fromString("50");
        var schemaXOnly = InitialPositionMapper.map(ipXOnly);
        assertThat(schemaXOnly).isNotNull();
        assertThat(schemaXOnly.getX()).isEqualTo(50.0);
        assertThat(schemaXOnly.getY()).isNull();
        assertThat(schemaXOnly.getDirection()).isNull();
    }
}
